src/Persistence/MassTransit.Azure.Storage/MessageData/AzureStorageMessageDataRepository.cs

```diff
@@ -0,0 +1,130 @@
+namespace MassTransit.Azure.Storage.MessageData
+{
+    using System;
+    using System.IO;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Context;
+    using MassTransit.MessageData;
+    using Microsoft.Azure.Storage;
+    using Microsoft.Azure.Storage.Auth;
+    using Microsoft.Azure.Storage.Blob;
+    using Util;
+
+
+    public class AzureStorageMessageDataRepository :
+        IMessageDataRepository,
+        IBusObserver
+    {
+        readonly StorageCredentials _credentials;
+        readonly IBlobNameGenerator _nameGenerator;
+        readonly Uri _containerUri;
+
+        public AzureStorageMessageDataRepository(Uri storageEndpoint, string containerName, StorageCredentials credentials, IBlobNameGenerator nameGenerator)
+        {
+            _credentials = credentials;
+            _nameGenerator = nameGenerator;
+
+            _containerUri = new Uri(storageEndpoint, containerName);
+        }
+
+        public async Task<Stream> Get(Uri address, CancellationToken cancellationToken = default)
+        {
+            var blob = new CloudBlockBlob(address, _credentials);
+            try
+            {
+                await blob.FetchAttributesAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (StorageException exception)
+            {
+                throw new MessageDataException($"MessageData content not found: {blob.Container.Name}/{blob.Name}", exception);
+            }
+
+            return await blob.OpenReadAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
@phatboyg 
Are you calling `FetchAttributesAsync` just to find out if the blob exists?

If so, I do not think this is necessary. Internally, `CloudBlockBlob.OpenReadAsync` does a call to `FetchAttributesAsync`.

I think you can simply return the stream inside the try block.

Source: https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.storage.blob.cloudblob.openreadasync?view=azure-dotnet-legacy